[HUDI-2752] The MOR DELETE block breaks the event time sequence of CDC

danny0405 opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contribute/how-to-contribute before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieAppendHandle.java:
```diff
@@ -475,12 +475,12 @@ private void writeToBuffer(HoodieRecord<T> record) {
     } else {
-      keysToDelete.add(record.getKey());
+      keysToDelete.add(DeleteKey.create(record.getKey(), record.getData().getOrderingVal()));
     }
```
Hello, @vinothchandar , can you take a look if you have time ? The only concern is that the new encoding/decoding breaks the compatibility. I have no good idea how to be compatible yet.

But i want to address that before the path, the handle may cause data lost. Comparing to compatibility, correctness is more important i think.

danny0405 replied:
Got you, thanks.

cuibo01 replied:
>I have no good idea how to be compatible yet.

How to Resolve Data Compatibility

danny0405 replied:
Have added a new version

xushiyan reviewed (commented):
@danny0405 IIUC, the problem originates from `OverwriteWithLatestAvroPayload#combineAndGetUpdateValue` does not honor ordering value. `DefaultHoodieRecordPayload` does. If that's the case, by switching default payload to `DefaultHoodieRecordPayload` would solve the problem? plus asking users to migrate the payload class. cc @nsivabalan

danny0405 commented:
> @danny0405 IIUC, the problem originates from `OverwriteWithLatestAvroPayload#combineAndGetUpdateValue` does not honor ordering value. `DefaultHoodieRecordPayload` does. If that's the case, by switching default payload to `DefaultHoodieRecordPayload` would solve the problem? plus asking users to migrate the payload class. cc @nsivabalan

No, the problem has no relationship with payload clazz, it is how the log handle encode/decode the delete messages.

xushiyan commented:
@danny0405 i see. the main problem now is the new DeleteKey breaks the serde with log block contains HoodieKey, correct? The compatibility issue is quite impactful. I noticed there is a version bit in reading the block `int version = dis.readInt()` in `getKeysToDelete()` which is not used. Now might it be make use of?

danny0405 commented:
> @danny0405 i see. the main problem now is the new DeleteKey breaks the serde with log block contains HoodieKey, correct? The compatibility issue is quite impactful. I noticed there is a version bit in reading the block `int version = dis.readInt()` in `getKeysToDelete()` which is not used. Now might it be make use of?

Don't know how to do that ? Can you give some help in a following PR, appreciate that if you can help with that.

danny0405 commented:
@hudi-bot run azure

nsivabalan commented:
@danny0405 : Can you update the desc of the PR.

@danny0405 @xushiyan : I looked at the jira linked and w/o getting into implementation details (I have not looked into the actual fix yet), but here is my understanding.
Problem of interest: 
```
Currently, the DELETE blocks are always written after the data blocks for one batch of data write, when there are INSERT/UPDATEs after the DELETE, the data would lost.

What i can thought of is that the DELETE block should at least keep the event time sequence for #preCombine with other record payloads.
```

My take here is, we can't fix this in a fool proof manner unless we plan to retain delete records forever in hudi. 

for eg:
If we get an insert in log blog1, delete in log block2 and insert in log block3, as per master, we return the record of interest as valid since we saw an insert after delete (even if delete record's preCombine value is greater than the insert record following it).

if I am not wrong, the patch is fixing this, so that we still consider the record of interest as deleted since among all 3 versions of the record, the delete record has the higher preCombine value. 

But what incase compaction kicked in and then we came across a newer insert w/ lower preCombine. During compaction, we will not carry over any deleted records. 

To illustrate.
insert in log block1, and delete in log block2.
and then compaction kicks in.
and now lets say we have a new insert for the same record w/ lower preCombine. 

Since compaction will not carry over any deleted records, the new record will be considered a new insert. which is in line even if compaction has not kicked in (as per master). 

So, are we proposing to retain all deleted records forever in this patch or are we fixing only the case where we maintain the strict ordering only until compaction kicks in.

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+      // should be deleted or be kept. The old record is kept only if the DELETE record has smaller ordering val.
+      // For same ordering values, uses the natural order.
+
+      HoodieRecord<? extends HoodieRecordPayload> oldRecord = records.get(key);
```
whats the guarantee that oldRecord will be present here? 
what incase we get a delete record for which actual insert is in the base file and not in any of the log blocks.

danny0405 replied:
Then we should perform a `#combineAndGetUpdateValue` and the merging would take care of the event time ordering, which is still correct.

vinothchandar replied:
i think we check for existence now, above?

danny0405 replied:
Yes, we did the check already.

nsivabalan commented:
I skimmed through the patch. 
Looks like, we are looking to fix only a specific scenario. 

insert record with preCombine value 100.
delete record with preCombine value 50.

So, as per master, the record will be deleted. 
and with this patch, we will not delete the record and honor the insert as valid record. 

let me know if my understanding is right. And whats the plan for other scenarios which I have called out earlier.

danny0405 commented:
> I skimmed through the patch. Looks like, we are looking to fix only a specific scenario.
> 
> insert record with preCombine value 100. delete record with preCombine value 50.
> 
> So, as per master, the record will be deleted. and with this patch, we will not delete the record and honor the insert as valid record.
> 
> let me know if my understanding is right. And whats the plan for other scenarios which I have called out earlier.

Yes, your understanding is correct.

> So, are we proposing to retain all deleted records forever in this patch or are we fixing only the case where we maintain the strict ordering only until compaction kicks in.

It not only fixes the compaction but also the MOR reader, with this path, we would have same event time ordering semantics for both COW table and MOR table. Say, we should always obey the event time ordering of the messages and keep consistent query result no matter what the input sequence is.

alexeykudinkin commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestMORDataSource.scala:
```diff
@@ -516,17 +516,14 @@ class TestMORDataSource extends HoodieClientTestBase {
-    // the result might differ depending on whether compaction was triggered or not(after record is deleted). In this
-    // test, no compaction is triggered and hence we see the record from base file.
-    checkAnswer((1, "a0", 10, 100, false))
+    // Ordering value will be honored, the delete record is considered as obsolete
```
This comment is misleading -- previous record isn't a deletion

alexeykudinkin commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestMORDataSource.scala:
```diff
@@ -516,17 +516,14 @@ class TestMORDataSource extends HoodieClientTestBase {
 
     writeData((1, "a0", 18, 96, false))
-    // Ideally, once a record is deleted, preCombine does not kick. So, any new record will be considered valid ignoring
-    // ordering val. But what happens ini hudi is, all records in log files are reconciled and then merged with base
```
What this test describes seems plainly an issue to me. 
Let's 
 - Create a ticket to address it 
 - Keep the test flow (either w/in this test or separately) to make sure we have a reference point when we get to fix it

danny0405 replied:
This test describes is just so weird because it tries to explain why the error/unexpected case happens, i don't think there is any necessity to keep it.

If you think the fix the hard to understand, let's make it clear in this PR.

danny0405 replied:
So what is the meaning to keep an explanation for an incorrect case ? Let's just fix that and the confusion is gone

vinothchandar replied:
@alexeykudinkin can you clarify the high level concern here? do you think the test is wrong (with the changes in the PR) or we need more cases or you unearthed some bug we still need to fix in the PR

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieDeleteBlock.java:
```diff
@@ -86,7 +86,7 @@ public HoodieDeleteBlock(Option<byte[]> content, FSDataInputStream inputStream,
         byte[] data = new byte[dataLength];
         dis.readFully(data);
-        this.keysToDelete = SerializationUtils.<HoodieKey[]>deserialize(data);
+        this.keysToDelete = SerializationUtils.<DeleteKey[]>deserialize(data);
```
This is going to be non-BWC. 

To avoid surprises i think we should do either of
 - Support both log-formats (adding new format version for deletion blocks)
 - Make sure we can enforce proper upgrading sequence (first compact, then upgrade)
 - Make it BWC

danny0405 replied:
1 & 3 is okey to me, @xushiyan would follow this in another PR then ~

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
-        hoodieKey.getPartitionPath(), getPayloadClassFQN()));
+  protected void processNextDeletedKey(DeleteKey deleteKey) {
+    String key = deleteKey.getRecordKey();
+    if (records.containsKey(key)) {
```
Why making 2 hash-table lookups, instead of 1?

danny0405 replied:
Sorry, can't get your point.

alexeykudinkin replied:
`containsKey` and then `get` are the 2 lookups i was referring to. Why not instead do just 1 lookup?
```
value = records.get(key);
if (value != null) { ... }
```

cuibo01 replied:
yes, reduce 1 lookup, line 140 is the same

vinothchandar replied:
+1 . Don't want to harp on this though.

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+      // Checks the ordering value does not equal to 0
+      // because we use 0 as the default value which means natural order
+      boolean choosePrev = !newOrderingVal.equals(0)
+          && oldOrderingVal.getClass() == newOrderingVal.getClass()
```
This should be an assertion rather than part of the conditional

danny0405 replied:
We can not make an assertion because there is a default value: 0 which is an integer.

danny0405 replied:
I get your point but default value as long does not solve the problem, a use can also use integer version number.

danny0405 replied:
The `preCombine` field value can be any type, of course can be an integer.

alexeykudinkin replied:
For those that might be interested in the conversation: class check here is necessary, since Java doesn't perform automatic down-casting and instead would simply throw `ClassCastException` if, for ex, you'd try to compare int and long via `compareTo`

danny0405 replied:
How do you know which one is int and which one is long?

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+
+      HoodieRecord<? extends HoodieRecordPayload> oldRecord = records.get(key);
+      Comparable oldOrderingVal = oldRecord.getData().getOrderingVal();
+      Comparable newOrderingVal = deleteKey.getOrderingVal();
```
Nit: `deletionOrderingVal` resolves the confusion

danny0405 replied:
I would prefer `newOrderingVal`.

vinothchandar replied:
IMO currentOrderingVal and deleteOrderingVal is clearer. Leave it to you.

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/common/model/DeleteKey.java:
```diff
@@ -0,0 +1,87 @@
+ * be dropped if there are intermediate deletions for the inputs
+ * (a new INSERT comes after a DELETE in one input batch).
+ */
+public class DeleteKey extends HoodieKey {
```
Conceptually, `orderingVal` is not part of the key, it is rather part of the payload. I think it would be better to actually approach it that way:

1. Introduce `DeletionRecordPayload` (carrying only value of `orderingVal`)
2. Introduce new `DeleteLogBlock` format version appropriately persisting/reading new payload type

This has following benefits: 
 - Takes care of backward-compatibility (BWC) out of the box
 - Make it compatible with Metadata Table also out of the box

danny0405 replied:
`DeletionRecordPayload` is good suggestion, for a new format i would rather left it to other committers.

alexeykudinkin replied:
> For BWC, i would expect for other contributors can address this in following PR.

I don't think we can treat BWC as optional here. What do you think @vinothchandar?

alexeykudinkin replied:
> I kind of agree that, but that's the re-design/abstraction of the payload functionality and can we address that in separate issue and don't make things too complex here.

I don't think `DeleteKey` is the right abstraction here for the reasons i've referred to above, of which primary ones are 

 - It's non BWC
 - It's introducing new way of deleting records that make COW and MOR diverge

danny0405 replied:
> It's non BWC
Already agreed with that.

> It's introducing new way of deleting records that make COW and MOR diverge
Can you just go over with the work flow in detail of COW and MOR combining process first ? The DELETE records encode/decode for MOR table is always there for efficiency. And i didn't introduce new diverge because it is designed there before. This PR only tag the old delete keys with version number and fix the event time semantics, 

if you want to address that the abstraction is not correct, please give a correct abstraction and let's discuss there and i'm glad to see that.

danny0405 replied:
> +1 on providing concrete suggestions to move forward.
> 
> @danny0405 The main contention here is that `DeleteKey` is not really a key. we need some cleanup where we pass the orderingVal from the engine API (flink level), all the way down the actual places where I/O happens. Also mandate orderingVal across all the payloads. I was hoping we would do that in the new merge API design. But that's getting delayed now.

Thanks, here is the case, i'm trying to explain the details again

When consuming the CDC source, flink would tag each record with a `RowKind` corresponding to the `HoodieOperation` field of our format.

When user opens the changelog mode, we do not use the DELETE block encode and treat each delete records as normal data payload in order to keep the deleted val to use for retraction in streaming. So the current code can work correctly.

But in the default upsert mode, say user do not want to keep the hoodie operations flags for streaming read, the bug/problem comes in: We use the DELETE block encoding and if there is disorder deletes in the merge handle batch, the data lost happens.

That's why i try to fix that in this PR then. Hope this can give the help for the ideas of designing new merging APIs.

alexeykudinkin replied:
> Can you just go over with the work flow in detail of COW and MOR combining process first ? The DELETE records encode/decode for MOR table is always there for efficiency. And i didn't introduce new diverge because it is designed there before. This PR only tag the old delete keys with version number and fix the event time semantics,

Of course I'm aware of the merging process in both COW and MOR. Efficiency of DELETE block is not in the fact that we just store keys for the records we're planning to delete, right? 

> And i didn't introduce new diverge because it is designed there before. This PR only tag the old delete keys with version number and fix the event time semantics,

I've miscommunicated that -- what i meant was that, adding `DeleteKey` makes it further diverge from the COW implementation.

> +1 on providing concrete suggestions to move forward.

I've outlined concrete suggestion in the very first comments -- to fill in `orderingVal` (as well as potentially additional payload in the future) in a proper `RecordPayload` casing, so that we store not only keys of the deleted records but also a payload with meta information (`orderingVal` for now).

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/model/DeleteKey.java:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  public static DeleteKey create(String recordKey, String partitionPath) {
+    return create(recordKey, partitionPath, 0);
```
I use 0 as default value because we also use that in the data payload, see constructor `OverwriteWithLatestAvroPayload(Option<GenericRecord> record)`, generally i think there is no much difference between `0` and `-1` because the version val should usually be positive in practice.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java:
```diff
@@ -114,4 +114,16 @@ default T preCombine(T oldValue, Properties properties) {
+
+  /**
+   * This method can be used to extract the ordering value of the payload for combining/merging,
+   * or 0 if no value is specified which means natural order.
```
It means the record comes later should be chosen.

danny0405 replied:
Yeah, your terminology is acccurate.

danny0405 replied:
+1 for `getCombineValue `.

danny0405 commented on hudi-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java:
```diff
@@ -1091,6 +1092,123 @@ public void testAvroLogRecordReaderWithDeleteAndRollback(ExternalSpillableMap.Di
+    final List<String> readKeys = new ArrayList<>(200);
+    final List<String> emptyPayloadKeys = new ArrayList<>();
+    scanner.forEach(s -> readKeys.add(s.getRecordKey()));
+    scanner.forEach(s -> {
```
No, the current code is more clear to me. Combining is not the gold rule in all of the places.

danny0405 replied:
I don't want to argue with code style anymore, will just ignore those comments.

vinothchandar commented:
Talking a pass!

guanziyue commented:
Happy to see the discussion of this problem, I would like to share our solution of this problem. We chose to abandon delete block mechanism and treat all records as mutation on possibly existing record. 
1. Remove the logic to handle '_hoodie_is_deleted' in getInsertValue method in payload so that this method always return a non-empty generic record. 
2. When we meet a delete record, return a record which has filed 'has _hoodie_is_deleted' is true and ordering value field. I think this may be similar to the tombstone mentioned by @alexeykudinkin above.
4. Store all records in data block in which some of them is actually delete marker record. 
5. Handle physical deletion in HoodieWriteHandle. We have already take into consider in combineAndGetUpdateValue. But we removed the logic in getInsertValue before. Add it back in HoodieMergeHandle/HoodieCreateHandle because it is time to do physical operation.

Pros: we reuse some existing logic and it should be backward-compatibility.
Cons: _hoodie_is_deleted is no longer an optional field in hudi. We have to treat it same as other meta fields in HoodieRecord. In addition to this, this method will waste little storage because _hoodie_is_deleted is stored in parquet and always false. Not sure if we can remove it or not.

Besides, I would like to share some more information which helped me find this problem. This problem has bothered another MVCC storage engine for a long time. [Hbase limitation](https://hbase.apache.org/book.html#_current_limitations). And they have a trade-off solution finally [HBASE-15968](https://issues.apache.org/jira/browse/HBASE-15968)](https://issues.apache.org/jira/browse/HBASE-15968) similar to the conclusion mentioned by @nsivabalan . Hope these are helping.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+          && oldOrderingVal.getClass() == newOrderingVal.getClass()
+          && oldOrderingVal.compareTo(newOrderingVal) > 0;
+      if (choosePrev) {
+        // The DELETE message is obsolete if the old message has greater orderingVal.
```
nit: consistent terminology. "if the old record has greater orderingVal"

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+      boolean choosePrev = !newOrderingVal.equals(0)
+          && oldOrderingVal.getClass() == newOrderingVal.getClass()
+          && oldOrderingVal.compareTo(newOrderingVal) > 0;
+      if (choosePrev) {
```
nit: consistent terminolog. old vs Prev.

vinothchandar reviewed (commented):
Left some comments that can hopefully help us converge. BWC needs to fixed for sure. 

I am also trying to read through the merging code to ensure there are n't edge cases. stay tuned.

vinothchandar commented:
Looked at the code, seems like we do queue up the log blocks in order and for DMS/Debezium payloads we pass the right ordering field value., extending `OverwriteWithLatestAvroPayload`. So it should all work. 

Whats the payload used for Flink CDC? cc @danny0405

danny0405 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/table/action/commit/FlinkWriteHelper.java:
```diff
@@ -105,7 +105,7 @@ public static FlinkWriteHelper newInstance() {
       // everything
       // so pick it from one of the records.
-      boolean choosePrev = data1.equals(reducedData);
+      boolean choosePrev = data1 == reducedData;
```
Currently the payload clazz does not have general `#equals` implementation. And we either returns the prev record or cur record, so the `==` is more efficient.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -411,7 +411,7 @@ private void processDataBlock(HoodieDataBlock dataBlock, Option<List<String>> ke
    *
    * @param key Deleted record key
    */
-  protected abstract void processNextDeletedKey(HoodieKey key);
```
+1

danny0405 commented:
> Looked at the code, seems like we do queue up the log blocks in order and for DMS/Debezium payloads we pass the right ordering field value., extending `OverwriteWithLatestAvroPayload`. So it should all work.
> 
> Whats the payload used for Flink CDC? cc @danny0405
> 
> The default payload.

I have addressed the review comments, please review again ~

What does the payload look like for DELETE debezium record  then ?

danny0405 commented:
@hudi-bot run azure

danny0405 commented:
@hudi-bot run azure

vinothchandar commented:
Will review again.

cuibo01 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+      // because we use 0 as the default value which means natural order
+      boolean choosePrev = !deleteOrderingVal.equals(0)
+          && curOrderingVal.getClass() == deleteOrderingVal.getClass()
+          && curOrderingVal.compareTo(deleteOrderingVal) > 0;
```
in retract stream, deleteOrderingVal and curOrderingVal are the same...
so the condition is `curOrderingVal.compareTo(deleteOrderingVal) >= 0` right？

cuibo01 replied:
2： COW has the same issue？

danny0405 replied:
1. We take the process time semantics for DELETE and INSERT has the same event time.

2. COW has no issue because the DELETE was removed physically when generating new parquet file and in the in-memory deduplication we take the event time into account with invoking `#preCombine`.

cuibo01 replied:
>We take the process time semantics if DELETE and INSERT has the same event time.

what does that mean? in retract stream, deleteOrderingVal and curOrderingVal are the same regardless of the job usage event time or process time.

danny0405 replied:
I guess there must be a primary key for hudi and there should be an ordering field, or we should totally use the process time sequence and the `DeleteRecord` should not be used, changelog mode is more suitable, this patch only fixes the default upsert behavior: the upstream should be a retract stream with event time.

danny0405 replied:
@vinothchandar Maybe we should really deprecate the delete block forever, imagine there are operations in one traction for a database:

INSERT DELETE INSERT (for a key, they all only have process time semantics, the linearizabiity)

The record is kept for a RDBMS but would lost if we do these operations in one commit in Hudi even if after this patch.

vinothchandar replied:
Whats the concrete proposal to redo this? Can you write it down high level. I ll try to read this again and get the context.

danny0405 replied:
The concrete proposal original was to solve the data lost when there are: `DELETE`, `INSERT` events in one transaction/commit of hoodie table, if the events have correct event time field.

What the patch can not solve is: when the events have same event time field(processing sequence), the data would still be lost.

vinothchandar replied:
>would lost if we do these operations in one commit in Hudi even if after this patch.

is that not a problem of how we define ordering field? are you taking about cases specific to Flink CDC?

danny0405 replied:
> are you taking about cases specific to Flink CDC

The Flink CDC output should have a event time field there, but now they are all processing time sequence.

For Debezium source it works well with this patch.

cuibo01 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java:
```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
+      // Checks the ordering value does not equal to 0
+      // because we use 0 as the default value which means natural order
+      boolean choosePrev = !deleteOrderingVal.equals(0)
+          && curOrderingVal.getClass() == deleteOrderingVal.getClass()
```
curOrderingVal.getClass().getName().equals(deleteOrderingVal.getClass().getName())

cuibo01 replied:
![image](https://user-images.githubusercontent.com/36881649/159112163-bf0e9804-e0a5-4665-9c8d-1a9783c2ff64.png)

vinothchandar replied:
+1. we can add a utiltiy to ReflectUtils or ClassUtils for the comparison

nsivabalan replied:
can you help me understand how this would pan out for records written with 0.10.0 or before. 
I would assume DeleteRecord will have 0 as ordering value, where as, an insert record will have proper ordering value, lets say 10. 

so, if we get an insert followed by delete, we will choose delete record as final value. 
and if we get a delete followed by an insert, we will choose insert record as the final value.

danny0405 replied:
Before the patch, both `DELETE` -> `INSERT`  and `INSERT` -> `DELETE` choose `DELETE`, with this patch, `DELETE` is only chosen if it has a bigger ordering value.

danny0405 replied:
For same JVM classloader, use `==` is more efficient BTW, i can move it to `ReflectionUtils`.

cuibo01 commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieDeleteBlock.java:
```diff
@@ -86,15 +88,25 @@ public HoodieDeleteBlock(Option<byte[]> content, FSDataInputStream inputStream,
+  private static DeleteRecord[] deserialize(int version, byte[] data) {
+    if (version == 1) {
+      // legacy version
+      HoodieKey[] keys = SerializationUtils.<HoodieKey[]>deserialize(data);
```
I think adding log to tell the user that there is an delete record of old version, it may cause data loss.

vinothchandar commented:
@guanziyue That's a solid approach, treating everything as upserts. 

In general, for Spark (with micro-batching), the ordering/precombining gets rid of this scenario and you don't end up with a missed insert. But when you go directly into a log, streaming mode, then the merge needs to factor in the ordering time.

vinothchandar commented:
> Maybe we should really deprecate the delete block forever, imagine there are operations in one traction for a database:

@danny0405 the problem is more about ordering the records correctly, based on a unique ordering field for each operation right? For your example of `I D I` with the same event time, we might make it `I I` (data block) and `D` (delete block). I think what you are saying is - we want to preserve the source ordering and say encode its as `I D I` in the same data block?

nsivabalan commented on hudi-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java:
```diff
@@ -1097,6 +1097,123 @@ public void testAvroLogRecordReaderWithDeleteAndRollback(ExternalSpillableMap.Di
+    header.put(HoodieLogBlock.HeaderMetadataType.INSTANT_TIME, "104");
+    HoodieDeleteBlock deleteBlock3 = new HoodieDeleteBlock(deletedRecords3.toArray(new DeleteRecord[0]), header);
+    writer.appendBlock(deleteBlock3);
+
```
can we add another data block following this deleted block, with higher ordering value compared to those records in deleteBlock3. may be 5 of them. And assert that, new inserts are honored. 

Also, add few more inserts w/ lower ordering value compared to those in deleteblock3 and ensure deletes take precedence.

danny0405 replied:
It is not that easy to construct update with same keys for sampling data, `testPreCombineFiledForReadMOR` already cover this.
